Modular Fault System Rupture Set/Solution (v2)

kevinmilner opened the pull request:
This is version 2 of a working draft of my idea for how to re-think the nasty hierarchy of FaultSystemRupSet/Solution classes. There was lots of good discussion (especially with respect to the motivation) in #21, which I will close shortly.

cc: @field-usgs @chrisbc @voj @pmpowers-usgs 

I have created a new system of loading, tracking, and storing modules. This will likely be useful outside of rupture sets & solutions, so it's located in the `org.opensha.commons.util.modules` package. Here's the structure:

## Module Interfaces

In their simplest form, a module is anything that implements the tagging interface _OpenSHA_Module_. There are also a lot of helper interfaces, here's the structure:

* _OpenSHA_Module_: basically just a tagging interface for a module, does nothing other than extend Named
  * _ArchivableModule_: adds methods for storing/loading module information to/from a zip archive
    * _FileBackedModule_: provides default implementations for the above, requiring that implementations just supply a file name and ability to write to/init from streams
      * _CSV_BackedModule_: same, but now to/from a CSV
      * _JSON_BackedModule_: same, but now to/from a JSON
      * _TextBackedModule_: same, but now to/from plain text
  * _SubModule_<E extends _ModuleContainer_<?>>: a module that knows about its container, and through generics can enforce that its container only be a certain type

## Module Containers

Modules can be stored in and retrieved from containers. The base class for this is **ModuleContainer** which provides `addModule(OpenSHA_Module)`, `hasModule(Class<? extends OpenSHA_Module)`, and `getModule(Class<? extends OpenSHA_Module)`.

There is a subclass that handles loading module collections to/from zip files: **ModuleArchive** extends **ModuleContainer**. Here, modules are loaded lazily by using `container.addAvailableModule(Callable<E> call, Class<E> clazz)`.

Something can be both a module and a container, in that case modules are nested in the zip file (see below). Further, if a module implements the _SubModule_<E extends _ModuleContainer_<?>> interface, it will be able to access an instance of its container, including after it is loaded from a file.

### Inheritance Considerations

The container handles inheritance, and so if you have multiple possible implementations of any module and register a subclass, it will be retrievable using it's subclass or using the common base class. Further, if you can only have one instance loaded at a time for any module tree (to prevent accidental loading of multiple implementations of the same module). For example, consider these classes:

```
	abstract class AbstractCustomModule implements OpenSHA_Module {
		// abstract module code
	}
	
	class CustomModuleImpl extends AbstractCustomModule {
		// concrete module code
	}
	
	class AltModuleImpl extends AbstractCustomModule {
		// alt module code
	}
```

If you add a module that is of type `CustomModuleImpl`, you can retrieve it via either calling `getModule(CustomModuleImpl.class)` or `getModule(AbstractCustomModule.class)`, as both the base class and concrete implementation are of type OpenSHA_Module and thus would be mapped to the registered instance. If you later add a module of type `AltModuleImpl`, it will first evict the `CustomModuleImpl` (as they both share the same parent module class, `AbstractCustomModule`).

Helper interfaces that should never be registered to a module can be tagged with the `@ModuleHelper`. This prevents things like the _CSV_BackedModule_ helper interface from ever being mapped to an implementation, and prevents unrelated modules from being evicted if you add two modules that use the same helper.

## Rupture Set & Solution Implementations

Rupture Sets and Solutions are both modules and containers for sub-modules. The basic structure is as follows, but note that there is no inheritance here:

* ModularArchive: archive that contains top level modules (rupture set and solution
  * FaultSystemRupSet: rupture set module
    * sub modules
  * FaultSystemSolution: solution module
    * sub modules

Here are the class definitions:

```
package org.opensha.sha.earthquake.faultSysSolution;

public class FaultSystemRupSet extends ModuleContainer<OpenSHA_Module> implements ArchivableModule,
  SubModule<ModuleArchive<OpenSHA_Module>> {
     // contents here
}

public class FaultSystemSolution extends ModuleContainer<OpenSHA_Module> implements ArchivableModule,
SubModule<ModuleArchive<OpenSHA_Module>> {
     // contents here
}
```

You load a rupture set or solution via static `load(File)` methods, and you write them via `getArchive().write(File)` methods. If you load an rupture set that has a solution attached, the solution will still be written out even if you never accessed the solution (or didn't even know it was there). These classes make all file I/O methods final, such that they can only be loaded and stored in their general form that will always be accessible (e.g., if you make a NZRupSet class it will be loaded back in as a regular FaultSystemRupSet, and thus be accessible in SCEC-VDO). All custom information should be added as modules.

I made the old rupSet/solution implementations extend these new classes, so it's a drop in replacement and we can slowly ween ourselves off of the old classes. This should make adoption less painful.

### File Structure

Here's what it looks like as a zip archive:

* _modules.json_: master module index, lists rupture set and solution modules
* _ruptures_: sub-directory for rupture information & sub-modules
  * _ruptures/modules.json_: rupture set sub-modules
  * _ruptures/section_indices.csv_: data file used by the rupture set module
  * _ruptures/properties.csv_: data file used by the rupture set module
  * _ruptures/info.txt_: data file used by the Info module that was added to the rupture set module
* _solution_: sub-directory for solution information & sub-modules
  * _solution/modules.json_: solution sub-modules
  * _solution/rates.csv_: data file used by the solution module
  * _solution/info.txt_: data file used by the Info module that was added to the solution module

...and here are the contents of those various module.json files:

modules.json:

```
[
  {
    "name": "Rupture Set",
    "className": "org.opensha.sha.earthquake.faultSysSolution.FaultSystemRupSet",
    "path": "ruptures/"
  },
  {
    "name": "Solution",
    "className": "scratch.UCERF3.FaultSystemSolution",
    "path": "solution/"
  }
]

```

ruptures/modules.json:

```
[
  {
    "name": "Info",
    "className": "org.opensha.sha.earthquake.faultSysSolution.modules.InfoModule",
    "path": "ruptures/"
  }
]
```

solution/modules.json:

```
[
  {
    "name": "Info",
    "className": "org.opensha.sha.earthquake.faultSysSolution.modules.InfoModule",
    "path": "solution/"
  }
]
```

## TODO list

* Deprecate old classes
* Finalize file formats for rupture sets. [Draft documentation here](https://github.com/opensha/opensha/wiki/Modular-Fault-System-Solution).

kevinmilner committed:
initial module (v2) implementation and tests

kevinmilner committed:
added sub module concept, made rupSet/sol extend new implementations

kevinmilner committed:
serialization methods now final

kevinmilner committed:
tests for sub modules

kevinmilner committed:
renamed writeArchive -> write

kevinmilner committed:
some new modules, cluster rupture builder now uses new classes

kevinmilner committed:
draft of new logic tree framework, not yet used

chrisbc commented on src/main/java/org/opensha/sha/earthquake/faultSysSolution/ruptures/util/SegmentationCalculator.java:
```diff
@@ -1407,8 +1408,8 @@ public static void main(String[] args) throws IOException, DocumentException {
 				1000d, RuptureConnectionSearch.CUMULATIVE_JUMPS_DEFAULT);
-		rupSet.buildClusterRups(rsConnSearch);
-		List<ClusterRupture> rups = rupSet.getClusterRuptures();
+		rupSet.addModule(ClusterRuptures.instance(rupSet, rsConnSearch));
```
@kevinmilner I think this is potentially problematic, but first my assumptions:

 - that line 1411-new ==  line 1410-old >>  to build the Cluster Ruptures list using some helper class
 - line 1412-new & line 1411-old copy these objects to the local member variable 'rups'

My key concern is that it's reasonably easy to see what is happening in the old API from the method names and introspection in IDE etc. This is a  good thing & for new users especially, as the domain is being 'explained' through the API.  However, in the new API it less clear, as we seem to rely on the module class names and the module 'concept'. It' almost like we're creating a 'language' with only two or three verbs, which I fear may be rather difficult to understand for casual/new users.

kevinmilner committed:
moved module helpers to own package, lots of new tests

kevinmilner committed:
simplified, removed most generics

kevinmilner committed:
can now get/init from JSON string

kevinmilner committed:
serialization bugfixes, starting to incorporate old branch

kevinmilner committed:
old logic tree now extends new

kevinmilner committed:
new modules from U3 code, truly use refactor to use new when possible

kevinmilner committed:
more refactoring, can almost run inversion w/ only new format

kevinmilner committed:
fixed bug in new requireModule method

kevinmilner committed:
added notes on what and how methods are used

kevinmilner committed:
XML serialization will now specify border type, if applicable

kevinmilner committed:
exposed setZonePolygon() method

kevinmilner committed:
minor: constant iteration order for floating point reproducibility

kevinmilner committed:
added check for iteration order, exposed array to get all

kevinmilner committed:
minor changes to be floating-point compatible with U3 implementation

kevinmilner committed:
no-arg constructor for deserialization

kevinmilner committed:
new rup set builder, tests to ensure consistency with U3

kevinmilner commented:
Lots of progress to report on. First I have implemented a new rupture set builder, including a method for reproducing a UCERF3 inversion rupture set:

```
FaultSystemRupSet.builder(subSects, rups).forU3Branch(branch);
```

I have tested all of the various arrays and such against the original UCERF3 implementations and verified that they match perfectly to full double precision [JUnit here](https://github.com/opensha/opensha/blob/modular-fault-sys-rup-set2/src/test/java/org/opensha/sha/earthquake/faultSysSolution/RupSetBuilderTests.java). Once I got things to match perfectly, I modified both codes a tiny bit to use FaultSection.getArea(boolean) instead of doing their own calculations, for simplicity. That method has a different multiplication order so it doesn't match the previous implementation perfectly to floating point precision, but I think it's better to use the standard methods.

I also modularized all of the inputs necessary for generating inversion inputs, and verified that the matrices match perfectly without using any of the old classes. The same is true after writing the new files to a (new style) archive, and loading them back in, with the caveat that I had to copy over the input fault section data (due to issue #25, same would occur with the old classes and I/O). Those tests are just in the main method of [UCERF3InversionInputGenerator](https://github.com/opensha/opensha/blob/modular-fault-sys-rup-set2/src/main/java/scratch/UCERF3/inversion/UCERF3InversionInputGenerator.java) for now as they require a lot of memory and aren't well suited for CI.

So I think we're in really good shape. @chrisbc  and @voj, I can show you how to configure inversions using the new code next week. Next steps include:

* Serializing grid source providers
* Improving fault section file format
* Generalizing the inversion setup code to be more flexible for UCERF4 (longer term goal)
* Deprecating the old code, and letting the new code load in the old file formats

kevinmilner committed:
cleaned up lots of imports mangled by the auto-refactor tool

kevinmilner committed:
Merge pull request #26 from opensha/master

Update modular branch from master

kevinmilner committed:
Merge pull request #28 from opensha/master

Catch up modular branch from master

kevinmilner committed:
Merge pull request #29 from opensha/master

Catch up modular branch with GeoJSON fault section

kevinmilner committed:
sect areas & slip rates are now modules w/ default impl

kevinmilner committed:
renamed to U3FaultSystemRupSet/Solution/IO

kevinmilner committed:
cleanup & deleted old one off/utility classes no longer used

kevinmilner committed:
initial (incomplete) serialization, TODO: region ser

kevinmilner committed:
preparing for new serialization

kevinmilner committed:
temp commmit to get around merge conflict

kevinmilner committed:
Merge pull request #31 from opensha/master

Catch up modular branch w/ GeoJSON serialization updates

kevinmilner committed:
grid source serialization

kevinmilner committed:
deprecated GridSourceFileReader, moved scale to super interface

kevinmilner committed:
convenience methods

kevinmilner committed:
demo files for docs (and future tests)

kevinmilner commented:
This is almost ready, and I now have a complete draft of the file formats for core modules (gridded seismicity was the last one).

I have posted a draft of file format documentation, which is intended mostly for outside users wishing to implement the models in other codes (it doesn't currently describe the module JSON format). As an outside-ish user, would you be willing to take a look @pmpowers-usgs?

Docs here: https://github.com/opensha/opensha/wiki/Modular-Fault-System-Solution

Demo files here: https://github.com/opensha/opensha/tree/modular-fault-sys-rup-set2/src/test/resources/org/opensha/sha/earthquake/faultSysSolution

I also renamed the old (in scratch) classes to U3FaultSystemRupSet and U3FaultSystemSolution, and deleted some unused one-off test classes.

The updated TODO list is:

* Add formal tests for rupture set/solution file formats, loading, default modules, etc
* Deprecate old classes
* Finalize file formats for rupture sets
* Add fallback code if you try to load the new objects from an old file

kevinmilner committed:
Merge pull request #32 from opensha/master

Add new location to modular branch

pmpowers-usgs commented:
Yes, but not for a few days

kevinmilner committed:
can now deserialize array and location properties by default

kevinmilner committed:
moved setRandom to superinterface

kevinmilner committed:
added ability to attempt to load an unlisted module from an archive

kevinmilner committed:
more robust FS[R]S loading, tests, bugfixes

kevinmilner committed:
rup set builder test requires too much memory, disabling

kevinmilner committed:
increased memory available to gradle tests

kevinmilner committed:
added new testVerbose task that prints tests as they complete

kevinmilner committed:
tests of standard modules & related serialization bugfixes

kevinmilner committed:
added test timing information

kevinmilner committed:
sped up slow tests

kevinmilner committed:
fixed U2 grid sources after location change

kevinmilner committed:
new individual rates module (for average solutions)

kevinmilner committed:
corrected model weights for final UCERF3 model

kevinmilner committed:
LogicTreeLevel now generic, added LogicTree class

kevinmilner committed:
fixed compile errors & U3 logic tree level names

kevinmilner committed:
fixed bug with enum unchecked method

kevinmilner committed:
new file format separates rupture section IDs and properties

kevinmilner committed:
minor file name changes

kevinmilner commented:
I updated the file format following feedback from @chrisbc to separate out the section indices for each rupture, and properties (like magnitude/rake) into separate CSV files. [Documentation has been updated here to reflect these changes](https://github.com/opensha/opensha/wiki/Modular-Fault-System-Solution).

kevinmilner committed:
added switch for temporarily enabling backwards-compatible locations

kevinmilner committed:
now includes standard module tests

kevinmilner committed:
extracted interfaces for fault-node polys/associations, U3 loads cached

kevinmilner committed:
added README to fault system rupture sets/solutions

kevinmilner committed:
test should now pass

kevinmilner committed:
minor GeoJSON improvements, e.g., default display property constants

kevinmilner committed:
Merge branch 'master' into modular-fault-sys-rup-set2

# Conflicts:
#	src/main/java/org/opensha/commons/geo/GriddedRegion.java
#	src/main/java/org/opensha/commons/geo/json/Feature.java
#	src/main/java/org/opensha/commons/geo/json/FeatureProperties.java
#	src/main/java/org/opensha/commons/geo/json/Geometry.java

kevinmilner committed:
fix for featureproperty location serialization

kevinmilner committed:
diagnostic page now uses new format

kevinmilner committed:
added NSHM23 v1.2 fault model data

kevinmilner committed:
moved rup set configuration from main to new RuptureSets class

kevinmilner committed:
added missing RuptureSets class

kevinmilner committed:
added default serialization for common function classes

kevinmilner committed:
extracted superclass/module for InversionTargerMFDs, added serialization

kevinmilner committed:
archive now lists assets for each module (not currently used)

kevinmilner committed:
new slip along rupture model class, enum defers to class

kevinmilner committed:
simplified new slip along rupture model configuration

kevinmilner committed:
fixed bug where dip direction not propagated to sub sections

kevinmilner committed:
new GeoJSON support

kevinmilner committed:
now writes GeoJSON for connections & histograms

kevinmilner committed:
dip direction now actually set correctly

kevinmilner committed:
neighbors connection strategy now done threaded as well

pmpowers-usgs commented:
Review comments for modular fault system solution:

**Documentation looks great!**

Solution comments:
- `README` --> markdown
- (nit) Don't put the trailing slash in `"path"` in `modules.json` files. Use `java.nio.file` internally to resolve paths.
- Reduce precision across the board on export of the solutions and fix floating point errors where present in magnitudes and lat-lons.

Ruptures comments:
- `section_indices.csv` --> `rupture_indices.csv` | `indices.csv`
  - Drop the `# 1,# 2, ...` part of the header in section_indices; the readme should explain what the additional columns are after the header
- `ave_slips.csv` --> `slips.csv` | `average_slips.csv`
  - Why isn't `slip` in `properties.csv`?
- `fault_sections.geojson` features need `DipDirection`
- `properties.csv` should include dip and width; reduce precision:
  - magnitude to 3 decimal places
  - rake to 0 or 1 decimal place; integer value is fine too
  - `area` and `length` to, say, 5 decimal places (6 sig figs); Is there a compelling reason to have area and length in `m` as opposed to `km`?

Solution comments:
- `grid_region.geojson`
  - I'd be wary of storing grid coordinates as a `MultiPoint` nested inside a `GeometryCollection` (I don't like these because there is no way to define properties for the component geometries). Consider WUS or AK at 0.1° spacing? 4x increase if we go to 0.05°.
  - I don't understand `LatNodes` and `LonNodes`
  - Consider using `grid_region.geojson` as follows:
    - a single `Polygon` feature that is the bounds of the grid
    - include `spacing` property; are lat and lon spacing ever different? To date, NSHMs are always the same and just have a single value
    - consider dropping `anchor`; the stored coordinates should suffice
    - use optional `bbox` geojson member to define mercator bounds of region

- `*_mfds.csv`
  - Fix floating point precision errors in magnitude headers
  - Consider also having `mfds.csv` that user knows are total gridded mfds
  - Reduce precision of mfd rates to 5 or 6 decimal places.


Good rounding method to clean up values to fixed precision when serializing:
``` java
  /**
   * Round a double to a specified number of decimal places according to the
   * supplied {@link RoundingMode}. Internally this method uses the scaling and
   * rounding capabilities of {@link BigDecimal}. Note that a negative
   * {@code scale} will round {@code value} to the specified number of places
   * above the decimal.
   *
   * @param value to round
   * @param scale the number of decimal places in the result
   */
  public static double round(double value, int scale, RoundingMode mode) {
    return BigDecimal.valueOf(value).setScale(scale, mode).doubleValue();
  }
```

kevinmilner commented:
Thanks a log @pmpowers-usgs, this is really helpful. A few comments/questions/ramblings that come to mind as I mull your suggestions over:

> Reduce precision across the board on export of the solutions and fix floating point errors where present in magnitudes and lat-lons.

I certainly agree that things would be more user friendly with more appropriate precision. The downside of doing so is that [de]serialized files wouldn't exactly match the rupture sets/solutions that were created in memory. So if you created a rupture set in memory and generated inputs for an inversion, those inputs would be slightly different if you had loaded in the same rupture set from a file. Any differences would likely be immaterial, however. I suppose we could do the rounding on construction rather than on serialization to keep things consistent.

> Why isn't slip in properties.csv?

I have gone back and forth on this one. The current thinking is that "average slip" is an optional property that is not required for hazard calculations, just for running inversions (satisfying slip rate constraints), and thus shouldn't be in the required `properties.csv` file. Externally built rupture sets might not have this property defined, and I don't bother with slips when I load in RSQSim catalogs as rupture sets (though the information is there in RSQSim).

> fault_sections.geojson features need DipDirection

Agreed, I actually ran into this yesterday and forced it to always be written from here on out.

> properties.csv should include dip and width

Interesting -- those things are calculated on the fly from the fault sections in OpenSHA, do you think it would be more convenient for users to have them included in the file? In that case I would write them out, but probably ignore them on load. I guess the same could be said for area and length, though lengths aren't always the sum of all subsection lengths (e.g., faults with multiple subsections down-dip).

> I'd be wary of storing grid coordinates as a MultiPoint nested inside a GeometryCollection

Do you propose a different way of storing them? It's very important to me to store the grid nodes rather than infer them, as I don't want to ever be in the position where a change in the Location class or gridding algorithms breaks compatibility with previous results (as occurred recently when we changed Location). I need to reliably be able to load in the exact same grid.

> I don't understand LatNodes and LonNodes

They're optional and can be included to ensure that any deserialized GriddedRegions are exactly match the original version, but can be ignored by most users.

> include spacing property; are lat and lon spacing ever different? To date, NSHMs are always the same and just have a single value

I'm mirroring the GriddedRegions class that supports different spacings, and don't see the harm. I can see it being used for extreme latitudes. You might consider it for Alaska where equal spacing in degrees would mean approximately double the horizontal resolution: at lat=65 N, 0.1 degrees of longitude is 5km whereas 0.1 degrees of latitude is still 11 km.

> use optional bbox geojson member to define mercator bounds of region

Can you elaborate on the usefulness of this?

pmpowers-usgs commented:
Good points. Further thoughts:

> dip and width

Most users aren't messing with individual ruptures. However, reconstructing average dips and widths from fault sections requires a bunch of steps, some of which I could see users getting wrong (e.g. the width when averaging is measured in the DipDir, not normal to the upper trace of a potentially trapezoidal subsection). I'd gotten used to having them available because they feed directly into GMMs, no muss, no fuss. However, I don't think we (the NSHMP) are parameterizing ruptures the way we should be given how the GMMs are developed and their sensitivity to things like hanging wall effects. Averaged values from multi-fault ruptures give rise to parameterizations that don't make sense or are inconsistent. I'd advocate continuing to export them for now with possible removal/deprecation at a later date.

> grid_region.geojson

As far as I can tell, you've got the actual border of the region encoded in `Lat|LonNodes` and the the rectangular (mercator) 'bounding box' (hence the `bbox` suggestion) encoded in the GC > Polygon feature. The nodes themselves exist in 4 files. In the NSHMs, we use geojson polygons for borders and other properties of the region, but that file is secondary to csv files which enumerate the specific nodes along with MFD properties or explicit rates in magnitude columns. I'd like to know where Ned is on all the bookkeeping business; the on-fault, off-fault, supra-, sub-seimogenic distinctions are a lot of complexity that  we don't _actually_ know is an improvement, or not, over dealing with the just the fault (and fault-like structures, or zones) in the inversion and then computing a final off-fault/background MFD. We're refactoring the NSHMs to use spatial PDFs in the model files. These can be used to scale a-values or an explicit MFD. Double-counting in grid cells then gets handled as the model loads, truncating any node MFD at the mMin of fault-based ruptures in the cell. (In UCERF3 there was then a step where we went back and renormalized the spatial PDF based on the truncations; very confusing and difficult to explain/remember). The only point of this ramble is that it would be much simpler to define solutions that contain a spatial PDF and the background MFD.

kevinmilner committed:
draft of new report framework

kevinmilner committed:
plot index support

kevinmilner committed:
new section-by-section conn pages, and fixed sect name bug w/ GeoJSON

kevinmilner committed:
added rounding by default to new rupture sets/solutions

kevinmilner committed:
fixed bug with sig-fig precision rounding

kevinmilner committed:
new demo files

kevinmilner committed:
fix for inset legend with log axis

kevinmilner committed:
solution MFD plots

kevinmilner committed:
generalized sect-by-sect pages, added solution plots to them

kevinmilner committed:
lots of report updates + full pipeline demo

kevinmilner committed:
new inv related modules, better plots, fixed slow demo SA speed

kevinmilner committed:
lots of SA updates: new averaging, better plots & debug, better perturb

kevinmilner committed:
better output & new exponential scale perturbation function

kevinmilner committed:
sped up threaded SA by only tracking indv constraints in outer thread

kevinmilner committed:
new rate vs rate scatter plot, inversion progress plot updates

voj committed:
spelling fixes

kevinmilner committed:
Merge pull request #37 from GNS-Science/fixes/spelling

spelling fixes

kevinmilner committed:
Merge branch 'modular-fault-sys-rup-set2' of git@github.com:opensha/opensha.git into modular-fault-sys-rup-set2

kevinmilner committed:
fixed file names after @voj's spelling fix (#37)

kevinmilner committed:
new 2D rate vs rate hist

kevinmilner committed:
can now set exponential perturbation orders of magnitude externally

kevinmilner committed:
renamed scratch resources dir, now ignored when building jars

kevinmilner committed:
more scrach data dir updates

kevinmilner committed:
slip rate calculation added to SARM

kevinmilner committed:
minor plot updates

kevinmilner committed:
now uses slip rate calc from SlipAlongRuptureModels

kevinmilner committed:
U3InversionTargertMFDs instance no longer needed to calc target slips

kevinmilner committed:
utility methods for non-equal-weight subplots

kevinmilner committed:
misc report updates

kevinmilner committed:
draft solution logic tree module

voj committed:
fixes #40: shutdown ExecutorService

kevinmilner committed:
Merge pull request #41 from GNS-Science/fix/tsa-shutdown

fixes #40: shutdown ExecutorService

kevinmilner committed:
Now uses Daemon threads for #40

kevinmilner committed:
core modules added via builder now lazily initialized

kevinmilner committed:
now supports new file formats

kevinmilner committed:
new branch averaged solutions & logic trees working

kevinmilner committed:
epistemic list ERF now works with sol logic tree, & optimizations

kevinmilner committed:
fixed bug with unassociated off fault MFD

kevinmilner committed:
incremental MFDs didn't have an offset, which make them spiky (see #43)

kevinmilner committed:
updated single-strand representation to use min-dist between clusters

kevinmilner committed:
OpenSHA defaults to depth-down in km in GeoJSON files. Assumes omitted
depths in fault traces should align with upper depth of surface

kevinmilner committed:
plot improvements: histograms on sect detail pages

kevinmilner committed:
now plots upper edge of fault surface in GeoJSON, not trace

kevinmilner committed:
new variable perturbation option

kevinmilner committed:
progress toward command line tools

kevinmilner committed:
added missing utility class

kevinmilner committed:
now always writes zero depths in fault traces for buried dipping faults

kevinmilner committed:
added ability to infer a gridded region from a node list

kevinmilner committed:
can now infer gridded reigon for grid seis if GeoJSON omitted

kevinmilner committed:
fixed bug on depth-detection for polygons

kevinmilner committed:
fixed NPE on energy calc without inequality constraints

kevinmilner committed:
now omits lines in MFD csv for nodes without MFDs

kevinmilner committed:
now assumes slip along rupture model is UNIFORM unless one is supplied

kevinmilner committed:
minor doc and reuse method instead of duplicating

kevinmilner committed:
added detection code to catch & correct region area non-singularity

kevinmilner committed:
Merge pull request #44 from opensha/region-nonsingularity-extra-path-fix

Fix for Regions created with non-singular Areas that contain an extra, empty path

kevinmilner committed:
new command line tool for sub-sectioning

kevinmilner committed:
crude rupture set CLI

kevinmilner committed:
bug with presets argument

kevinmilner committed:
fault system CLI tools updates

kevinmilner committed:
changed order of constraints in simple azimuthal for efficiency

kevinmilner committed:
fixed bug in total azimuth filter with only 1 end section

kevinmilner committed:
now uses cmd parsing from utility class

kevinmilner committed:
now prunes empty areas recursively, fixes @voj's 2nd edge case

kevinmilner committed:
added note about recursion

chrisbc committed:
fix for null pointer exeception when d_ineq == null;

kevinmilner committed:
now uses shared static method for row calculation (can be used in #45)

kevinmilner committed:
now detects gzipped input/output files by extension

kevinmilner committed:
added default GSON TypeAdapter

kevinmilner committed:
convenience methods for centered text

kevinmilner committed:
promoted class to calculate total FSS moment to base class

kevinmilner committed:
will now work (& disable a-priori constraint) if FM == null

kevinmilner committed:
now works with basic LogicTreeBranch (provided it has correct vals)

kevinmilner committed:
convenience method to calc deformation model moment rate

kevinmilner committed:
new total rate inversion constraint

kevinmilner committed:
minor plot updates

kevinmilner committed:
better auto-range in energy plots

kevinmilner committed:
new command line interface for simple inversions

kevinmilner committed:
now tries to detect terminal width when printing help

kevinmilner committed:
fix for #42, and better purterbation tracking with averaging

kevinmilner committed:
improved CLI

voj committed:
Added filled polygons to plots

kevinmilner committed:
disabled total moment constraint because it isn't helpful

kevinmilner committed:
improved table headings

kevinmilner committed:
Merge pull request #49 from opensha/fix/calc_energy_with_no_ineq

fix for null pointer exception when d_ineq == null;

kevinmilner committed:
now will accept lower case presets in CLI

kevinmilner committed:
Merge branch 'master' into modular-fault-sys-rup-set2

kevinmilner committed:
getModules() now defaults to load available modules

kevinmilner committed:
new geologic deformation model (v1)

kevinmilner committed:
improved print output

kevinmilner committed:
better y axis ranges

kevinmilner committed:
removed debug print

kevinmilner committed:
added ugly catch for magic numbers in deformation files...yuck

kevinmilner committed:
participation and slip rate plot improvements

voj committed:
added setYofX to XY_DataSet

voj committed:
added MFD_WeightedInversionConstraint

voj committed:
added MFDUncertaintyWeightedInversionConstraint

kevinmilner committed:
fix for y-range if constraint has zeros

kevinmilner committed:
Merge pull request #45 from opensha/feature/uncertainty_weighted_mfd_constraint

added new MFD Uncertainty weighted constraint;

kevinmilner committed:
Merge pull request #50 from GNS-Science/oakley/plot-polys

Added filled polygons to plots

kevinmilner committed:
Marked new polygon-fill type as not compatible with symbols (see #50)

kevinmilner committed:
CLI can now use attached InversionTargetMFDs

kevinmilner commented:
I think the time has finally come to merge this in, as development on master has ceased, things seem to be in working order, and file formats are now relatively stable. Will also merge in related branches on dev, cybershake, and SCEC-VDO repos.
